garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+    Abstract class for models with automatic pickling.
+
+    This class follows the BaseModel API and handles pickling automatically.
+    It provides both self.init_spec and self.parameters for pickling, which
+    are the arguments and weights of the Model. When unpickled, a new model
+    will be reconstructed with self.init_spec as arguments and load the weights
+    from self.parameters.
+
+    The target weights should be assigned to self._default_parameters during
+    pickling, so that the newly created model can check if target weights exist
+    or not.
+    """
+
+    def __init__(self):
+        self._name = type(self).__name__  # name default to class
+        self._inputs = None
+        self._outputs = None
+        self._default_parameters = None
+
+    @property
+    def init_spec(self):
+        """
+        Model specification.
+
+        Positional and named arguments (args and kwargs) which, when passed
+        to `__init__`, will construct this model.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        return self._inputs[0]
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        return self._outputs[0]
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        return self._inputs
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        return self._outputs
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        raise NotImplementedError
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        args, kwargs = self.init_spec
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
+
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
+        out._default_parameters = data['parameters']
+        self.__dict__.update(out.__dict__)
+
+
+class TfModel(AutoPickableModel):
+    r"""
+    Model class for TensorFlow.
+
+    This class is a parent-child design.
+    The parent model contains the parameters, i.e. the structure, and
+    two child model are constructed with input_1 and input_2. Inputs and
+    outputs of the parent model will be binded with the default child, which
+    is the first child constructed.
+
+    When new child model is created, it reuses the parameter from the
+    parent model and can be accessed by calling model.child_2, where
+    child_2 is the name given when building this particular child model.
+    If a child is constructed without given a name, the name "default" will
+    be used.
```

Comment:
* Note that each can only be built once
* Use the verb build/built, since "construct" already has a very specific meaning in Python